src/swtpm/mainloop.c

```diff
@@ -102,6 +102,7 @@ int mainLoop(struct mainLoopParams *mlp,
     struct iovec        iov[3];
     uint32_t            ack = htobe32(0);
     struct tpm2_resp_prefix respprefix;
+    uint32_t            lastCommand;
```

Comment:
is this state being migrated?

Reply:
No.

Reply:
Sending a TPM2_Shutdown() twice before a TPMLIB_Terminate(), will not affect the state of the TPM 2. Actually the spec says:

https://trustedcomputinggroup.org/wp-content/uploads/TCG_TPM2_r1p64_Part3_Commands_15may2021.pdf

"This command is used to prepare the TPM for a power cycle. The shutdownType parameter indicates
how the subsequent TPM2_Startup() will be processed.

[...]
This command saves TPM state but does not change the state other than the internal indication that the
context has been saved. The TPM shall continue to accept commands. If a subsequent command
changes TPM state saved by this command, then the effect of this command is nullified. The TPM MAY
nullify this command for any subsequent command rather than check whether the command changed
state saved by this command. If this command is nullified. and if no TPM2_Shutdown() occurs before the
next TPM2_Startup(), then the next TPM2_Startup() shall be TPM2_Startup(CLEAR)."

Reply:
> is this state being migrated?

Phew, actually that could in one cornercase become a problem. If the machine migrates with last command sent TPM2_Shutdown(SU_STATE)... swtpm writes not state at all today but only libtpms.

Reply:
... but then on the other hand the TPM2_Shutdown() that this patch here is sending only happens after the mirgation occurred, so it shouldn't affect it at all.

Reply:
@elmarco VMs ins power management mode don't go down iirc, they just sit there and do nothing until one presses a key or so, right?

Reply:
Power management is a complex topic ;). But  VMs are generally simply suspended  or "paused" (not using PM).

The trouble is that if the VM is "live" migrated (saved & restored without PM suspend etc), it will lose the lastCommand state... We probably need another GET_STATEBLOB type for swtpm data.

Reply:
> "Just how often is this going to happen?" live migration? it depends. save/restore are common operations. Some centers balance the load with live migration too.

Live migration itself is not the problem. Live migration and then abrupt termination of the VM would be the issue. Any normal OS is typically supposed to be shut down and the TPM 2 just happens to require that the OS be properly shut down.

Reply:
Well then, do we actually need to do anything? DA lockout might just be working as intended.

Reply:
A one-line solution for this problem maybe here:

https://github.com/stefanberger/libtpms/blob/9165880bf72c26154b4b2aa354c89cb86a5e7a7a/src/tpm2/DA.c#L136-L150
```
    // If DA has not been disabled and the previous shutdown is not orderly
    // failedTries is not already at its maximum then increment 'failedTries'
    if(gp.recoveryTime != 0
       && gp.failedTries < gp.maxTries
       && !IS_ORDERLY(g_prevOrderlyState))
	{
#if USE_DA_USED
	    gp.failedTries += g_daUsed;
	    g_daUsed = FALSE;
#else
	    gp.failedTries++;
#endif
	    // Record the change to NV
	    NV_SYNC_PERSISTENT(failedTries);
	}
```

We could just comment the line `gp.failedTries += g_daUsed;`. That should have the same effect.

Reply:
> As for this series, it addresses the issue on the swtpm level for most scenarios except for migration and then abrupt termination case where no further command was sent to the TPM by the VM. I think in this case I would not go through the extra mile of migrating the lastCommand to the destination but let the TPM 2 specs kick in demanding proper shutdown.

If that's what the spec demands, then why do anything at all? let's not add different behaviours when the maching is migrated or not, please :)

Reply:
I would say this PR should address 99% of the cases where an abrupt shutdown happens. If someone wants to trigger the exception then they migrate a VM multiple times and terminate it every time to run into that one case that is not covered.

Reply:
We should leave a FIXME for migration support.

Reply:
Migration doesn't imply suspend-to-ram.

Reply:
Correct, it doesn't. 

When doing suspend-to-ram with EDK2 the last-sent command is unfortunately not a TPM2_Shutdown() but a TPM2_GetRandom(). In fact EDK2 could legitimately send yet other commands than the TPM2_GetRandom() *after* the TPM2_Shutdown() without affecting the correct functioning of TPM2_Startup(SE_STATE) upon resume. So tracking the last-send command can only be used as an optimization.
Upon resume from suspend-to-ram swtpm needs to handle a CMD_INIT (due to reset) which looks like the abrupt reset case where we need to send a TPM2_Shutdown() [original purpose of this patch]. Now in this case with EDK2 (or any other firmware) we may not have seen TPM2_Shutdown(SU_STATE) as the last command but that TPM2_GetRandom() instead. To not mess up the TPM 2's state for proper resume with TPM2_Startup(SU_STATE), swtpm needs to send a TPM2_Shutdown(SU_STATE) to the TPM 2, which we obviously do for the original reason to avoid DA lockouts. So this is where the *SU_STATE* versus *SU_CLEAR* in the TPM2_Shutdown(SU_STATE) comes from.